### ISSUE TITLE
sort_custom() giving error and (heavy) lag...

### ISSUE THREAD

**Issue by Qws** [godotengine/godot#22833]
2 days ago I decided to update/recompile 3.1 Godot.

Now sort_custom() is giving me an error and unplayable heavy lag because I'm using it in _process(delta)

The error message:

> ERROR: Array::sort_custom: Condition ' !p_obj->has_method(p_function) ' is true. returned: *this
>    At: core\array.cpp:261

I checked the array.cpp line 261 and it was updated 3 days ago: https://github.com/godotengine/godot/commit/6415454581c3ba0025da6b9bae42e060fa4e1508

As someone who is not a C++ programmer I've no idea how to work around it. Did the behavior of sort_custom() changed or is it a bug?

And if it's a bug, is it save to recompile Godot again without the newly added Array.cpp's 261 line?

**Godot version:**

Godot 3.1 alpha (non-official), commit 3bcab46ea513aea363951ec7301e80e262408970

**OS/device including version:**
Windows 10

**Issue description:**

I expect sort_custom array to just do its thing... now I'm unable to stop the lagging and error in the Array.cpp (line 261)

**Steps to reproduce:**
Use in GDscript the sort_custom() function and it will give an error. Pointing towards the Array.cpp file.

**Comment by DavidSichma** [comment]
Judging from the error I would say that you pass a nonexistent function to sort_custom. Please make sure you didn't make a typo.

But this does not explain the lag. Maybe some other commit is causing it. 
Or you try to sort and/or print a very big array every frame.
But without code I can not do more than to guess.

**Comment by akien-mga** [comment]
CC @DualMatrix

**Comment by Qws** [comment]
@DavidSichma 
No typo involved, because the function name is "sort", so, not so much room for possible misspell.

Did you tested the sort_custom() function in the latest build? No errors for you? If so, how?

For me sort_custom() worked fine in my project since Godot 2.1.4-ish, 3.0, and early 3.1. No error, no lags while still using same similar size arrays.

Only after recompiling Godot from latest master branch 2days ago it stopped working properly, getting the specific error and massive lag inside _process(delta). Haven't touched sort_custom logic.

**Comment by DavidSichma** [comment]
Can you please send me the script to take a closer look? Because sort_custom works fine for me.

**Comment by Qws** [comment]
@DavidSichma 
Here is just very simple usage of sort_custom().

It works fine in previous versions of Godot, but not recent commit.

```
extends Node2D

func _ready():
	var y= []
	y.insert(0, 2222)
	y.insert(1, 3333)
	y.sort_custom(MyCustomSorter, "lol")
	pass

class MyCustomSorter:
	static func lol(a, b):
		print("ok doing something")
		if(a > b):
			return true
		else:
			return false
		pass
```

**Comment by DavidSichma** [comment]
I think I located the issue.

You are passing an inner class with a static function. Commit 6415454 added a check to make sure the object you pass actually has this method. But it turns out that `has_method()` doesn't actually work for classes and static methods.

`has_method()` will always return `false` in this case resulting in the test failing although it should just work.

This should be a seperate issue.

**Comment by DavidSichma** [comment]
For now you could try to use non static functions and instance the class or adding the method to the main class and passing `self`. Or, if you really want to do it, remove the error checking line.

**Comment by Qws** [comment]
@DavidSichma 
I don't use the static function in my original code, but both non-static and static giving me the same error.

what do you actually mean by adding to the main class?

```
extends Node2D

func _ready():
	var y= []
	y.insert(0, 2222)
	y.insert(1, 3333)
	y.sort_custom(self, "lol")
	pass

func lol(a, b):
     print("ok doing something")
		if(a > b):
			return true
		else:
			return false
		pass
```

**Comment by DavidSichma** [comment]
Yes, this is what I meant.

Just fix the the indent and it should work.

**Comment by pawilon** [comment]
I just encountered the same issue. The workaround of instancing the inner class and not using static classes works, however the docs suggest a static approach (http://docs.godotengine.org/en/latest/classes/class_array.html#class-array-sort-custom), therefore it's confusing for the users.

I'd say revert https://github.com/godotengine/godot/pull/22737/commits/6415454581c3ba0025da6b9bae42e060fa4e1508 for now and reopen the original issue so that a better solution can be made.

CC @akien-mga who merged the original PR.

**Comment by DualMatrix** [comment]
The true issue seems to be the ```has_method``` function being wrong. So as soon as that is fixed the error will dissapear and according to @DavidSichma in #22838 the fix for that is trivial.

**Comment by DavidSichma** [comment]
I have barely tested my fix. It is a miracle at best this even works. I also just found out that this fix is not free of bugs.

**Comment by akien-mga** [comment]
I've reverted 6415454 for now with c730957, as the fix for #22838 doesn't seem that trivial in the end. The patch could be readded (as it was correct per se) once we have a solution for #22838.
